### ISSUE TITLE
Single-client behavior

### ISSUE THREAD

**Issue by mbronckers** [mbronckers/GI-PVI#4]
This issue serves to document the model's inference behavior under a single-client setting.

**Comment by mbronckers** [comment]
**Default settings**: (see config.py) in [experiment logs](https://github.com/mbronckers/GI-PVI/tree/main/results/2022-06-13-21.03.50_std-500epochs):

```
class Config:
    name: str = ""
    seed: int = 0
    plot: bool = True
    
    epochs: int = 1000
    
    N: int = 100        # Number of training data pts
    M: int = 10         # Number of inducing points
    S: int = 2          # Number of training weight samples
    I: int = 10         # Number of inference samples

    nz_init: float = 1e-3
    lr: float = 1e-2
    ll_var: float = 1e-3
    
    batch_size: int = 100

    prior: Prior = Prior.NealPrior
    dgp: DGP = DGP.ober_regression

    random_z: bool = False
    bias: bool = True

    dims = [1, 50, 50, 1]
```

Left is 100 epochs, right 500 epochs.

<img width="1575" alt="image" src="https://user-images.githubusercontent.com/25270093/173445461-a1878ef7-ed45-4663-a87e-ef6d2561d225.png">

Left is StandardPrior, right NealPrior. Both 500 epochs.
<img width="1575" alt="image" src="https://user-images.githubusercontent.com/25270093/173446303-58c6e9a2-3615-4f42-a1e6-f68b0e4d79ee.png">


**Observations**:
- NealPrior overfits / too confident
- StandardPrior underfits / not confident enough.

**Comment by mbronckers** [comment]
Here is another example of the Standard N(0,1) prior underfitting. The mean prediction is great, but the confidence in the seen training regions is too high

10k epochs, 100 inference samples.

<img width="690" alt="image" src="https://user-images.githubusercontent.com/25270093/173448360-89d6caad-b4d4-431e-a8fb-c46ce1ecacb8.png">

**Comment by mbronckers** [comment]
Ober's paper plots the +/- 2 std devs. In accordance, I now plot 5-95th percentile.

**Comment by mbronckers** [comment]
Neal Prior, 1000 epochs. Batch size=N=100. M=10

<img width="700" alt="image" src="https://user-images.githubusercontent.com/25270093/173452456-78a151c0-dea4-46db-9b7c-81390ba3b0c6.png">
<img width="700" alt="image" src="https://user-images.githubusercontent.com/25270093/173452592-207786cb-2473-426c-927f-af103beacc81.png">

**Observations**:
- This looks more like it; the uncertainty in the middle is more pronounced (noticeable in the 5-95 bounds)
- The plot across the entire domain reflects the increased uncertainty on the left & right. 
- The mean prediction outside the training data is in agreement with Ober's plot (see below).
- The only concern is that the uncertainty in the middle in not pronounced.

**Ober et al: **
![image](https://user-images.githubusercontent.com/25270093/173452697-78c1c1b0-61a2-4ff2-be8e-b128fbea2183.png)

**Comment by mbronckers** [comment]
**Config details**: N=1000, M=100, NealPrior, 1000 epochs

<img width="1635" alt="image" src="https://user-images.githubusercontent.com/25270093/173547121-20058c45-b778-4688-b4f9-4a3e5ce3f5ca.png">

- When you increase the number of data points and inducing points, the model fit + inference are as expected with appropriate variance in previously unseen domain space

**Comment by mbronckers** [comment]
Some interesting details on Ober’s toy regression experiment settings:
- Random inducing point initialization, N = 40, M = 100, 5000 epochs
- Learning rate 1-e2, full batch GD
- I=10 training weight samples,  S=100 inference weight samples
- Parameter (weight) scaling to ensure same scale for Adam optimizer (see appendix section H); they use a scaling factor of 3
- Log precision initialized to -4 for all but the last layer, where it’s 0

5000 epochs
<img width="833" alt="image" src="https://user-images.githubusercontent.com/25270093/173560244-189eff75-8844-418d-b216-99cef116c103.png">

Inducing init and after 5K epochs:
<img width="1693" alt="image" src="https://user-images.githubusercontent.com/25270093/173562316-8c8e0b77-3c06-4171-92b2-185adbfc21b1.png">

**Observations**:
- Using most of these settings, I am starting to get somewhat similar predictions. However, the one thing that doesn’t match up is the variance — which might be attributable to the prior, learning rate or, precision parameter. Still got to figure that one out.
- The inducing points move in a very interesting shape (read: likely incorrect)

**Comment by mbronckers** [comment]
Results using same configuration as above, but with Standard Prior N(0, 1)

Predictions
<img width="801" alt="image" src="https://user-images.githubusercontent.com/25270093/173566049-6d9357b2-929f-4967-89f5-d3199d6f40db.png">

Inducing points
<img width="1657" alt="image" src="https://user-images.githubusercontent.com/25270093/173566190-bf5f95f8-9d7c-4edd-8eaa-b30aadbe3563.png">

**Observations**:
- Underfitting / underconfident
- Inducing points do not move in weird pattern like with NealPrior, but this might be due to relatively smaller updates due to the KL prior regularization

**Comment by mbronckers** [comment]
Having fixed the sampling in [commit](https://github.com/mbronckers/GI-PVI/commit/9e21e95d71a18321be5c93eac2861f851d1ad9e6#diff-dca556e51cb441e6761b67224dbcf1b6a00a6e9e79540c2da39aee15aaa6624b), this issue now appears to be resolved. The model behavior is as expected under regular hyperparameters. [See results here](https://github.com/mbronckers/GI-PVI/tree/main/results/2022-06-15-20.38.40_fixed_it%3F).

Changing the sample line to `sample = self.mean + B.mm(B.cholesky(self.var), noise)` fixed the issue.

Using `self.mean` instead of `B.mm(B.pd_inv(self.prec), self.lam)` also saved 50% per sample call.
```
0:00:00.000013  # if self.mean was cached
0:00:00.002346 # self.mean fresh computation
0:00:00.005192  # B.mm(B.pd_inv(self.prec), self.lam)
```
I verified that it was never the case that the cached mean was out of date with the current parameters by running the below in a baseline run of 1000 epochs:
```
    def mean(self):
        """column vector: Mean."""
        if self._mean is None:
            self._mean = B.cholsolve(B.chol(self.prec), self.lam)
        else:
            np.testing.assert_allclose(B.to_numpy(self._mean), B.to_numpy(B.cholsolve(B.chol(self.prec), self.lam)))
        return self._mean
```

**Comment by mbronckers** [comment]
# Post-identical results

**Comment by mbronckers** [comment]
### Random_z initialization

- Machine: term6

<img width="999" alt="image" src="https://user-images.githubusercontent.com/25270093/177300458-056ff026-f594-4af4-b4f8-614fae5784b2.png">

<img width="1002" alt="image" src="https://user-images.githubusercontent.com/25270093/177300423-ff411cec-3071-4bd1-9c66-fff1383f731d.png">

<img width="1000" alt="image" src="https://user-images.githubusercontent.com/25270093/177300387-1d63cae6-0738-4962-ae22-40fffcf76958.png">

**Comparison v. Baseline-term6**
<img width="1616" alt="image" src="https://user-images.githubusercontent.com/25270093/177300660-9c8a1ff3-441c-4145-a7ac-5584726e14dd.png">

**Comment by mbronckers** [comment]
## Final layer yz init to 1

- Machine: term6

L (baseline), R (yz_init-1)

<img width="1613" alt="image" src="https://user-images.githubusercontent.com/25270093/177302371-e2e52258-9f23-485e-8bf4-b2056f0569b3.png">

<img width="1619" alt="image" src="https://user-images.githubusercontent.com/25270093/177302502-f2f613e2-439b-497b-a011-20ff299d68ba.png">

<img width="1608" alt="image" src="https://user-images.githubusercontent.com/25270093/177302550-1d8836cc-db60-43da-bf9b-b23231cbf213.png">
